File: Quick_Sort/Quick_Sort.py
# Quick sort in python

def quick_sort(arr):
    # Base case: if the list has 1 or fewer elements, it's already sorted
    if len(arr) <= 1:
        return arr
    
    # Selecting the pivot element (here, we choose the middle element)
    pivot = arr[len(arr) // 2]
    
    # Dividing the list into three parts: elements less than pivot, equal to pivot, and greater than pivot
    left = [x for x in arr if x < pivot]      # Elements smaller than the pivot
    middle = [x for x in arr if x == pivot]   # Elements equal to the pivot
    right = [x for x in arr if x > pivot]     # Elements greater than the pivot

    # Recursively sorting the left and right partitions
    return quick_sort(left) + middle + quick_sort(right)



data=[-2,45,0,11,-9]
print("Sorted Array in Ascending Order: ")
print(quick_sort(data))


# This code is contributed by Erfan Shafiee

File: Quick_Sort/Quick_Sort.cs
// Quick sort in C#
using System;

// function to swap elements
static void Swap(ref int a, ref int b)
{
    int t = a;
    a = b;
    b = t;
}

// function to print the array
static void PrintArray(int[] array, int size)
{
    for (int i = 0; i < size; i++)
        Console.Write(array[i] + " ");
    Console.WriteLine();
}

// function to rearrange array (find the partition point)
static int Partition(int[] array, int low, int high)
{
    // select the rightmost element as pivot
    int pivot = array[high];

    // pointer for greater element
    int i = (low - 1);

    // traverse each element of the array
    // compare them with the pivot
    for (int j = low; j < high; j++)
    {
        if (array[j] <= pivot)
        {
            // if element smaller than pivot is found
            // swap it with the greater element pointed by i
            i++;

            // swap element at i with element at j
            Swap(ref array[i], ref array[j]);
        }
    }

    // swap pivot with the greater element at i
    Swap(ref array[i + 1], ref array[high]);

    // return the partition point
    return (i + 1);
}

static void QuickSort(int[] array, int low, int high)
{
    if (low < high)
    {
        // find the pivot element such that
        // elements smaller than pivot are on left of pivot
        // elements greater than pivot are on righ of pivot
        int pi = Partition(array, low, high);

        // recursive call on the left of pivot
        QuickSort(array, low, pi - 1);

        // recursive call on the right of pivot
        QuickSort(array, pi + 1, high);
    }
}

static void Main(string[] args)
{
    int[] data = { 8, 7, 6, 1, 0, 9, 2 };
    int n = data.Length;

    Console.WriteLine("Unsorted Array: ");
    PrintArray(data, n);

    // perform quicksort on data
    QuickSort(data, 0, n - 1);

    Console.WriteLine("Sorted array in ascending order: ");
    PrintArray(data, n);
}

// This code is contributed by Ali Barzegari Dahaj